File: core/src/com/inkostilation/pong/engine/Ball.java
package com.inkostilation.pong.engine;

import com.inkostilation.pong.engine.geometry.Circle;
import com.inkostilation.pong.engine.geometry.Rectangle;

import java.util.Random;

import static java.lang.Math.PI;

public class Ball extends Circle implements IUpdatable {

    private static final float MAXBOUNCEANGLE = (float) (5*PI/12);

    private float velocity = 80f;
    private float xVel, yVel, defaultX, defaultY;


    public enum Direction {
        LEFT(-1), RIGHT(1);

        public final int value;

        Direction(int value) {
            this.value = value;
        }
    }

    public Ball(float x, float y) {
        super(x,y, 10);
        defaultX = x;
        defaultY = y;
        this.xVel = velocity * getRandomDirection();
        this.yVel = velocity * getRandomDirection();
    }

    public void resetPosition() {
        setX(defaultX);
        setY(defaultY);
    }

    public void resetVelocity(Direction direction) {
        yVel = 0;
        xVel = velocity * direction.value;
    }

    private int getRandomDirection() {
        if (new Random().nextBoolean())
            return 1;
        else
            return -1;
    }

    public void applyCollision(Rectangle rectangle) {
        float relativeIntersectY = rectangle.getY() + rectangle.getHeight() / 2 - getY();
        float normalizedRelativeIntersectionY = 2 * relativeIntersectY / rectangle.getHeight();
        float bounceAngle = normalizedRelativeIntersectionY * MAXBOUNCEANGLE;

        int newDir = xVel > 0? -1 : 1;

        xVel = (float) (newDir * velocity * Math.cos(bounceAngle));
        yVel = (float) (-velocity * Math.sin(bounceAngle));
    }

    @Override
    public void update(float delta) {
        setX(getX() + xVel * delta);
        setY(getY() + yVel * delta);
    }

    public void constrain(Rectangle limit)
    {
        yVel = -yVel;

        if (getY() - getRadius() < limit.getY()) {
            setY(limit.getY() + getRadius());
        } else if (getY() + getRadius() > limit.getY() + limit.getHeight()) {
            setY(limit.getY() + limit.getHeight() - getRadius());
        }

        if (getX() - getRadius() < limit.getX()) {
            setX(limit.getX() + getRadius());
        } else if (getX() + getRadius() > limit.getX() + limit.getWidth()) {
            setX(limit.getX() + limit.getHeight() - getRadius());
        }
    }

    public void setVelocity(float velocity) {
        this.velocity = velocity;
    }
}

File: core/src/com/inkostilation/pong/engine/Ball.kt
package com.inkostilation.pong.engine

import com.inkostilation.pong.engine.geometry.Circle
import com.inkostilation.pong.engine.geometry.Rectangle
import java.util.*

class Ball(private val defaultX: Float, private val defaultY: Float) : Circle(defaultX, defaultY, 10f), IUpdatable {
    private var velocity = 80f
    private var xVel: Float
    private var yVel: Float

    enum class Direction(val value: Int) {
        LEFT(-1), RIGHT(1);

    }

    fun resetPosition() {
        x = defaultX
        y = defaultY
    }

    fun resetVelocity(direction: Direction) {
        yVel = 0f
        xVel = velocity * direction.value
    }

    private val randomDirection: Int
        private get() = if (Random().nextBoolean()) 1 else -1

    fun applyCollision(rectangle: Rectangle) {
        val relativeIntersectY = rectangle.y + rectangle.height / 2 - y
        val normalizedRelativeIntersectionY = 2 * relativeIntersectY / rectangle.height
        val bounceAngle = normalizedRelativeIntersectionY * MAXBOUNCEANGLE
        val newDir = if (xVel > 0) -1 else 1
        xVel = (newDir * velocity * Math.cos(bounceAngle.toDouble())).toFloat()
        yVel = (-velocity * Math.sin(bounceAngle.toDouble())).toFloat()
    }

    override fun update(delta: Float) {
        x = x + xVel * delta
        y = y + yVel * delta
    }

    fun constrain(limit: Rectangle) {
        yVel = -yVel
        if (y - radius < limit.y) {
            y = limit.y + radius
        } else if (y + radius > limit.y + limit.height) {
            y = limit.y + limit.height - radius
        }
        if (x - radius < limit.x) {
            x = limit.x + radius
        } else if (x + radius > limit.x + limit.width) {
            x = limit.x + limit.height - radius
        }
    }

    fun setVelocity(velocity: Float) {
        this.velocity = velocity
    }

    companion object {
        private const val MAXBOUNCEANGLE = (5 * Math.PI / 12).toFloat()
    }

    init {
        xVel = velocity * randomDirection
        yVel = velocity * randomDirection
    }
}